### PR TITLE
RFC: implement GraphQLConfig as the proposed protocol suggests

### PR DESCRIPTION
**Pull request by asiandrummer**
https://github.com/graphcool/graphql-config/issues/20 proposes a GraphQL configuration protocol - this PR implements the said protocol and improves inline documentations

WIP - submitting a PR for an early feedback. I will include a test suite soon.

### CODE REVIEW

File: src/config/GraphQLConfig.js [hunk 1]
```diff
@@ -59,109 +64,143 @@ export async function getGraphQLConfig(configDir: Uri): Promise<GraphQLRC> {
...
+  getExcludeDirs(appName: ?string): Array<Uri> {
+    if (appName) {
+      if (
+        this._config[APP_EXTENSIONS_NAME] &&
+        this._config[APP_EXTENSIONS_NAME][appName] &&
+        this._config[APP_EXTENSIONS_NAME][appName].excludeDirs
+      ) {
+        return this._config[APP_EXTENSIONS_NAME][appName].excludeDirs;
+      }
     }
```
Comment (wincent) [inline src/config/GraphQLConfig.js]: Probably worth extracting this pattern into a separate function.

```
getConfig<T: any>(keyName: string, appName: ?string, defaultValue: T = []): T {
  if (
    appName &&
    this._config[APP_EXTENSIONS_NAME] &&
    this._config[APP_EXTENSIONS_NAME][appName] &&
    this._config[APP_EXTENSIONS_NAME][appName][keyName]
  ) {
    return this._config[APP_EXTENSIONS_NAME][appName][keyName];
  }
  return this._config[keyName] || defaultValue;
}
```

And then using it to implement `getExcludeDirs` and `getIncludeDirs` etc.

Comment (wincent) [inline src/config/GraphQLConfig.js]: Or something like this (you get the idea).

File: src/config/GraphQLConfigTypes.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ *  LICENSE file in the root directory of this source tree.
+ *
+ *  @flow
+ */
+
+export type GraphQLConfiguration = {
+  schemaPath?: string,  // may be .json or .graphql
+  schemaUrl?: string,
+  // If env is specified, use one of the app configs in here
+  env?: GraphQLConfigurationEnvs,
```
Comment (wincent) [inline src/config/GraphQLConfigTypes.js]: Support for this seems unimplemented now, so should we leave this out for now?

File: src/config/__tests__/GraphQLConfig-test.js [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    const config = await getGraphQLConfig(CONFIG_DIR);
+    expect(config instanceof GraphQLConfig).to.equal(true);
+  });
+});
+
+describe('GraphQLConfig', () => {
+  let config;
+  beforeEach(async () => {
+    config = await getGraphQLConfig(CONFIG_DIR);
+  });
```
Comment (wincent) [inline src/config/__tests__/GraphQLConfig-test.js]: `beforeEach` here because you are planning on adding more than one `it`?

**Comment by wincent** [comment]
Just a heads up that once https://github.com/graphql/graphql-language-service/pull/47 lands this file will have moved from `src/config/GraphQLConfig.js` to [`packages/graphql-language-service-config/src/index.js`](https://github.com/graphql/graphql-language-service/blob/master/packages/graphql-language-service-config/src/index.js).

It's just a straight move with no other changes at this time, so hopefully should be easy to rebase this onto it just by copying it into the new place. I'm still working on the set-up of these packages, though: the package has no tests (and no test package dependencies), so I'll need to sort that out in order for the tests added in this PR to copy over cleanly.

File: src/config/GraphQLConfigTypes.js (hunk 2 above)
Comment (asiandrummer) [inline src/config/GraphQLConfigTypes.js]: This also describes the configuration protocol, so I think it'd be good to have it even without the implementation

**Comment by asiandrummer** [comment]
After attending to the comments, I've merged this commit in several different commits due to the merge conflicts && package version bump constraints. Closing this one in favor.

**Pull request closed without merging**